contributors/design-proposals/storage/container-storage-interface-cluster-registry.md

```diff
@@ -0,0 +1,158 @@
+    // configuration for it.
+	Spec CSIDriverSpec `json:"spec"`
+
+	// Status of the CSI volume driver.
+	Status CSIDriverStatus `json:"status,omitempty"`
+}
+
+// CSIDriverList is a collection of CSIDriver objects.
+type CSIDriverList struct {
+    metav1.TypeMeta `json:",inline"`
+    
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is the list of CSIDrivers
+	Items []CSIDriver `json:"items"`
+}
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    // Driver indicates the name of the CSI driver that this object refers to.
+    // This MUST be the same name returned by the CSI GetPluginName() call for
+    // that driver.
+    Driver string `json:"driver"`
+    
+    // Indicates this CSI volume driver requires an attach operation (because it
+    // implements the CSI ControllerPublishVolume() method), and that Kubernetes
+    // should call attach and wait for any attach operation to complete before
+    // proceeding to mounting.
+    // If value is not specified, default is true -- meaning attach will be
+    // called.
+    // +optional
+    AttachRequired *bool `json:"attachRequired"`
+    
+    // Indicates this CSI volume driver requires additional pod information
+    // (like podName, podUID, etc.) during mount operations.
+    // If this is set to true, Kubelet will pass pod information as
+    // VolumeAttributes in the CSI NodePublishVolume() calls.
+    // If value is not specified, default is false -- meaning pod information
+    // will not be passed on mount. 
+    // +optional
+    PodInfoRequiredOnMount *bool `json:"podInfoRequiredOnMount"`
+}
+
+// CSIDriverStatus is the status of a CSIDriver.
+type CSIDriverStatus struct {
+    // Indicates the volume driver has been successfully deployed on the cluster
+    // and is ready to use.
+	Ready bool `json:"ready"`
+}
+
+```
+
+#### CRD Installation
+
+The `CSIDriver` object schema will checked in to kubernetes/kubernetes repo under a storage directory.
+The `CSIDriver` `CustomResourceDefinition` (CRD) will be installed by a new Kubernetes controller that is responsible for ensuring required CRDs are installed.
+The controller will periodically verify that the CRD is still installed, and recreate it, if it is not.
+
+#### CR Creation
+
+When a CSI volume driver is deployed on Kubernetes, it may optionally, register it self with Kubernetes by creating a new `CSIDriver` object.
+The Kubernetes team will modify the CSI [`driver-registrar` container](https://github.com/kubernetes-csi/driver-registrar) to automatically do this.
+If kubelet driver registration is enabled, the kubelet will automatically do this as part of plugin registration.
+The driver must set the `Driver` field to the same name returned by the CSI `GetPluginName()` call for that driver.
```

Comment:
So does this mean that a CSI driver has to ship with some mechanism of instantiating CRs when it is deployed? Currently obviously that is not part of CSI spec, so how will that get done?

Reply:
> So does this mean that a CSI driver has to ship with some mechanism of instantiating CRs when it is deployed?

CSI driver authors can choose to but don't have to

> Currently obviously that is not part of CSI spec, so how will that get done?

The same way all custom k8s behavior is done, side car containers provided by us (e.g. driver-registrar, external-attacher, etc.).